Implements changes to allow developers more control about event handling

TheCataliasTNT2k opened the pull request:
Adds Framework::dispatch_automatically
Makes client::dispatch::update_cache_with_event public, so it can be called manually by developers

GnomedDev committed:
Remove temporary `Vec` usage in the `http` module (#2624, #2646)

This avoids having to allocate to store fixed length (replaced with normal
array) or fixed capacity (replaced with `ArrayVec`) collections as vectors for
the purposes of putting them through the `Request` plumbing.

Slight behavioral change - before, setting `params` to `Some(vec![])`
would still append a question mark to the end of the url. Now, we check
if the params array `is_empty` instead of `is_some`, so the question
mark won't be appended if the params list is empty.

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

GnomedDev committed:
Remove deprecated items (#2645)

GnomedDev committed:
Remove more deprecated items (#2651)

GnomedDev committed:
Remove `*_arc` methods (#2654)

These are unnecessary. Accepting `impl Into<Arc<T>>` allows passing either `T` or `Arc<T>`.

GnomedDev committed:
Clean up `ShardManager`/`ShardQueuer`/`ShardRunner` (#2653)

GnomedDev committed:
Put `Message::thread` behind a `Box` (#2658)

This trades a heap allocation for messages sent along with thread
creation for `Message`'s inline size dropping from 1176 bytes to 760
bytes,

GnomedDev committed:
Replace `Vec` and `String` with `FixedArray` and `FixedString` for all models (#2656)

This shrinks type sizes by a lot; however, it makes the user experience slightly
different:

- `FixedString` must be converted to String with `.into()` or `.into_string()`
  before it can be pushed to, but dereferences to `&str` as is.
- `FixedArray` must be converted to `Vec` with `.into()` or `.into_vec()`
  before it can be pushed to, but dereferences to `&[T]` as is.

The crate of these types is currently a Git dependency, but this is fine for
the `next` branch. It needs some basic testing, which Serenity is perfect for,
before a release will be made to crates.io.

GnomedDev committed:
Fix truncation when using `FixedString<u8>` where multibyte characters push over the limit (#2660)

Kerollmops committed:
Encode timestamps in RFC3339 (#2665)

Fixes #2664

ARandomDev99 committed:
Switch to `i64` in methods associated with `CreateCommandOption` (#2668)

This commit:

- switches from `u64` to `i64` in `CreateCommandOption::min_int_value` and
`CreateCommandOption::max_int_value` to accommodate negative integers in
Discord's integer range (between -2^53 and 2^53). Values outside this
range will cause Discord's API to return an error.
- switches from `i32` to `i64` in `CreateCommandOption::add_int_choice` and
`CreateCommandOption::add_int_choice_localized` to accommodate Discord's
complete integer range (between -2^53 and 2^53). Values outside this
range will cause Discord's API to return an error.

GnomedDev committed:
Get rid of the duplicate users cache (#2662)

This cache was just duplicating information already present in `Guild::members`
and therefore should be removed.

This saves around 700 MBs for my bot (pre-`FixedString`).

This has to refactor `utils::content_safe` to always take a `Guild` instead
of`Cache`, but in practice it was mostly pulling from the guild cache anyway
and this means it is more likely to respect nicknames and other information,
while losing the ability to clean mentions from DMs, which do not matter.

GnomedDev committed:
Change `Embed::fields` to use `FixedArray` (#2674)

`Embed::fields` previously had to stay as a `Vec` due to `CreateEmbed` wrapping
around it, but by implementing `Serialize` manually we can overwrite the
`Embed::fields` with a normal `Vec`, for a small performance hit on
serialization while saving some space for all stored `Embed`s.

GnomedDev committed:
Use `FixedArray` and `FixedString` in model enums (#2675)

Simply missed these when finding and replacing.

GnomedDev committed:
Bitpack boolean fields using fancy proc macro (#2673)

This uses the `bool_to_bitflags` macro to remove boolean (and optional boolean)
fields from structs and pack them into a bitflags invocation, so a struct with
many bools will only use one or two bytes, instead of a byte per bool as is.

This requires using getters and setters for the boolean fields, which changes
user experience and is hard to document, which is a significant downside, but
is such a nice change and will just become more and more efficient as time goes
on.

GnomedDev committed:
Use `nonmax::NonMax*` types in `Option`s wherever possible (#2681)

This swaps fields that store `Option<Int>` for `Option<NonMaxInt>` where the
maximum value would be ludicrous. Since `nonmax` uses `NonZero` internally,
this gives us niche optimisations, so model sizes can drop some more.

I have had to include a workaround for [#17] in `optional_string` by making my
own `TryFrom<u64>`, so that should be removable once that issue is fixed.

[#17]: https://github.com/LPGhatguy/nonmax/issues/17

GnomedDev committed:
Remove unused warning `#[allow(...)]`s (#2682)

A couple of clippy bugs have been fixed and I have shrunk model
sizes enough to make `clippy::large_enum_variant` go away.

GnomedDev committed:
Get rid of `unsafe` (#2686)

A discord bot library should not be using the tools reserved for low
level OS interaction/data structure libraries.

GnomedDev committed:
Swap Git dependencies for crates.io published versions (#2685)

GnomedDev committed:
Swap Id from `NonZero` to `NonMax` (#2689)

Discord seems to internally default Ids to 0, which is a bug whenever
exposed, but this makes ID parsing more resilient. I also took the
liberty to remove the `From<NonZero*>` implementations, to prevent future
headaches, as it was impossible to not break public API as we exposed
`NonZero` in `*Id::parse`.

GnomedDev committed:
Rewrite builders to take `Cow`s (#2688)

GnomedDev committed:
Store `Request::params` as a slice instead of owning `ArrayVec` (#2694)

This,
1. shrinks the size of Request, when copied around, as it doesn't have
to store the max capacity at all times
2. shrinks llvm-lines (compile time metric) for my bot in debug from
`1,153,519` to `1,131,480` as no monomorphisation has to be performed
for `MAX_PARAMS`.

mkrasnitski committed:
Revert `Request::params` back to `Option` (#2695)

Follow-up to #2694.

When `Request::params` was made into an ArrayVec, the `Option` around it
was removed in order to avoid having to add a turbofish on `None` to
specify the value of `MAX_PARAMS`. Also, `Request::new` also needed to
be changed so that the value of `MAX_PARAMS` could be inferred. Now that
the field is a slice again, we can wrap it in `Option` again (at no cost
to size, thanks to niche opts).

We ensure we never store the redundant `Some(&[])` by checking for an
empty slice and storing `None` instead. This way, we ensure we never
read an empty slice out of the `Some` variant.

GnomedDev committed:
Remove unnecessary `Vec` in `WsClient` (#2705)

GnomedDev committed:
Feature gate `#[instrument]` macros  (#2707)

The instrument macros generate 2% of Serenity's release mode llvm-lines,
and are proc-macros so hurt compile time in that way, so this limits
them to opt-in. This commit also fixes the issues that the instrument macro
was hiding, such as results that didn't ever error and missing
documentation.

GnomedDev committed:
Remove `impl Into<Option>` (#2701)

This signature is hard to use as `None` cannot infer the type of the
generic. I also replaced `Option<u8>` with `Option<NonMaxU8>` as it's
more efficient and will make the user think of the maximum value.

GnomedDev committed:
Get rid of `IntoIterator` generics wherever they are inefficent (#2698)

This removes inefficient `IntoIterator` generics and instead takes what is
actually required. I also reworked `reorder_channels` to allow for keeping the
generic, as it actually does only just need iterator semantics.

GnomedDev committed:
Clean up `Ratelimiter` (#2687)

Previously, someone assumed that `Ratelimiter` was going to be cloned, so
put a ton of `Arc`s everywhere. This was unneeded, and before dashmap,
so the buckets were also stored massively inefficiently. This fixes all
that.

I had to shuffle around the `Ratelimit` methods a little bit to return
their sleep time instead of sleeping themselves, so I didn't have to
hold a dashmap lock over an `.await`.

GnomedDev committed:
Handle overflow checks in central locations (#2697)

This removes multiple error variants and overall cleans up the codebase
by moving overflow checks into two `ModelError` variants.

GnomedDev committed:
Shrink size and clean up `Error` (#2700)

Shrinks `size_of::<Error>` from 136 bytes to 64 bytes, while removing unused
variants. This will improve performance for any method returning
`Result<T>` where `T` is less than the size of `Error` as both `Result`'s
`Ok` and `Err` have to be allocated stack space.

GnomedDev committed:
Remove manual `#[inline]` attributes (#2702)

The compiler knows best as inlining is quite complicated. This should
help with compile times, significantly.

GnomedDev committed:
Remove `Into<*Id>` and `AsRef<str>` (#2704)

`Into<*Id>` is often entirely unnecessary, and at worst hides the type of
the Id with hard-coded Ids. This fixes this by just using the Id type
itself, which also helps with monomorphisation bloat.

I also removed `AsRef<str>`, as this can just be `&str` without
requiring a generic and without requiring a call to as_ref in the
function, while also not accidentally taking ownership of a `String`
passed in misleadingly.

mkrasnitski committed:
Remove `Event::Unknown` variant (#2708)

If deserialization fails, we should allow the specific deserialization error to
be logged so that it can be reported to the Serenity maintainers. As it stands,
any events that fail to deserialize normally will "succeed" when deserialized
into the `Unknown` variant, throwing away any extra information about the
error.

Because deserialization errors are immediately logged when they happen,
explicitly whitelist `Error::Json` in `Shard::handle_event` to prevent
duplicate logs.

GnomedDev committed:
Use `NonMax` to signal builder range constraints (#2703)

GnomedDev committed:
Stop using `Value` in error parsing (#2710)

`Value` is bad to use and is slow. I also used `Arc<str>` to store `path` as
this is cloned a couple times.

GnomedDev committed:
Fix new clippy lints (#2713)

GnomedDev committed:
Update to `small-fixed-array` v0.2 (#2711)

`small-fixed-array` v0.2 removes implicit truncation and makes all
truncations explicit.

I also changed FixedString initialization from literals to use
`from_str_trunc` to take advantage of SSO if possible.

GnomedDev committed:
Fix `@everyone` role deserialisation (#2716)

The `@everyone` role sometimes has a role position of -1, but not always!

GnomedDev committed:
Fix `GuildChannel` deserialisation with `user_limit` (#2715)

This is documented as max 99 for voice channels, but 10k for stage
channels.

Milo123459 committed:
Fix maximum description length to the correct size (#2718)

GnomedDev committed:
Remove dashboard example (#2714)

This example depended on rillrate, which has been archived for over a
year.

arqunis committed:
Fix compilation from rebase

GnomedDev committed:
Remove deprecated `absolute_ratelimits` feature (#2719)

mkrasnitski committed:
Implement `max_concurrency` support when starting shards (#2661)

When a session is first kicked off and we need to start many shards, we
can use the value of `max_concurrency` to kick off multiple shards in
parallel.

GnomedDev committed:
Remove typemap without making everything generic (#2720)

Replaces the typemap with `dyn Any`, which is what typemap used
internally anyway. This gives the user more control over locking,
removes boilerplate, and is less confusing for beginners to learn.

GnomedDev committed:
Provide `Arc<Data>` instead of `&Data` (#2722)

This incurs an extra ref count, but this should be negligible for the
more permissive type. An explanation of how this works is that:

- `Arc` is just a shared pointer to `ArcInner`
- `ArcInner` stores the concrete Data type
- Each `Arc` stores the dyn vtable ptr
- `Arc` checks the vtable ptr to get the type id, then it can just give
out a new `Arc` referring to the `ArcInner` concretely.

GnomedDev committed:
Add `Client::try_data` to fallibly fetch the data type (#2723)

This is mostly for usage in `poise` to provide a nicer error message in
its situation. I also changed `Client::data` to provide `Arc` as well.

GnomedDev committed:
Fix voice manager initialisation (#2727)

mkrasnitski committed:
Remove some pedantic lints from the whitelist (#2728)

This also removes the impls of `From<{i32,u64}> for Colour`, as it only
makes sense to convert infallibly from `u32`.

GnomedDev committed:
Reduce unknown events to debug level, again (#2730)

arqunis committed:
Fix compilation from rebase

mkrasnitski committed:
Fix testing example

mkrasnitski committed:
Remove `PartialEq` implementations from all builders (#2734)

Because builder fields are all private, it doesn't make sense to compare
them directly. From a user's perspective, they have no idea what the
builder might contain. This also lets us change their internals easier,
without worrying about equality.

GnomedDev committed:
Remove various instances of `impl AsRef<T>` (#2736)

This should improve compile times even more, by compiling a lot more
methods in Serenity, instead of the user's code.

GnomedDev committed:
Remove `simd-json` support (#2735)

`simd-json` is not a performance improvement for Serenity.

It has a ton of sketchy uses of `unsafe`, increases the maintenance burden, and
leads to Serenity containing a json compatibility layer(!).

This commit removes all that, although a faster json decoding implementation is
on the table in the future if the compatibility layer is kept out of tree and
it is measurably faster for Serenity's usecase.

GnomedDev committed:
Fix massive code size of `client::dispatch` (#2739)

This replaces fancy iter stuff with just writing the code, and therefore
reduces the code size significantly and spawns less tasks.

GnomedDev committed:
Fix RoleTags deserialisation after removal of simd-json (#2742)

There was a comment here that said it was called by simd-json, so I
removed the function, but it turns out that serde_json also calls it.

GnomedDev committed:
Avoid cloning FullEvent in dispatch code (#2740)

Avoids cloning entire event structs when dispatching. To reduce code
size and allocations all the state is bundled up into a temporary struct
called DispatchContext, but this isn't user facing at all. This
regresses code size a bit, but is worth it for runtime performance, and
probably reduces actual code sizes as `FullEvent` doesn't have to be
moved around as much.

GnomedDev committed:
Remove the builder trait and only take required arguments (#2741)

Removes unnecessary trait, which only limited the signatures to always
take `CacheHttp` and required the async_trait boxing and compile time
cost. This also makes a lot less methods generic, improving compile
times.

GnomedDev committed:
Use Serializer::collect_seq in serialize impls (#2744)

GnomedDev committed:
Convert enum number into wrapper struct to save type sizes (#2746)

Saves a couple bytes in each model by replacing the enums with a simpler
and more accurate representation, a newtype with named statics.
https://www.diffchecker.com/ohzayL9Z/

GnomedDev committed:
Use Arc<str> to store the token (#2745)

`SecretString` is nice to have, but this is more efficient, and
`SecretString` is defense in depth we can do without.

GnomedDev committed:
Remove standard framework (#2731)

This PR carries out the removal after the deprecation made on current.

GnomedDev committed:
Re-add Secret to bot token (#2748)

This re-adds Secret to the bot token, without losing the `Arc<str>` opt.
This works by using a newtype to Zeroise on last drop, using
`Arc::get_mut`.

GnomedDev committed:
Take Arc<Http> in ClientBuilder::new_with_http (#2749)

This allows the user to make a http, pass it off to a background task,
then use it in serenity, instead of having to initialize two.

GnomedDev committed:
Remove duplicate content_safe implementation (#2759)

GnomedDev committed:
Simplify the message cache (#2757)

GnomedDev committed:
Upgrade to small-fixed-array 0.4 (#2755)

GnomedDev committed:
Avoid HashMap in ShardQueue (#2750)

vicky5124 committed:
Switch to a generic type argument for `ClientBuilder::data` (#2765)

GnomedDev committed:
Fix `VoiceServerUpdate::guild_id` being wrapped in `Option` (#2766)

https://discord.com/developers/docs/topics/gateway-events#voice-server-update

GnomedDev committed:
Improve MessageBuilder to fit with other builders (#2767)

GnomedDev committed:
Fill message cache with ChannelId::messages and ChannelId::pins (#2768)

GnomedDev committed:
Defer deserialisation of Dispatch to handle errors (#2764)

mkrasnitski committed:
Remove `CreateAttachment::id` field (#2771)

GnomedDev committed:
Fix remaining clippy warnings

arqunis committed:
Fix compilation from rebase

mkrasnitski committed:
Minor adjustments

arqunis committed:
Format unformatted code from rebase

arqunis committed:
Remove rebase artifacts in README

arqunis committed:
Synchronise README example from the `current` branch

cheesycod committed:
Apply Strum derives for FullEvent and Event (#2788)

GnomedDev committed:
Remove `Channel::name` (#2787)

`Channel::name` just called `to_channel` (cloning the entire `Channel`!) and
`PrivateChannel::name` or `GuildChannel::name`.

`GuildChannel::name` just returned the field of the same name, for consistency
with `PrivateChannel::name`.

`PrivateChannel::name` returned a non-standard format string that would be very
easy to replicate in user code.

I also tweaked the signature of `OnlineStatus::name` to stop passing an 8 byte
reference to a 1 byte struct, and return `&'static str`.

GnomedDev committed:
Use join_to_string throughout serenity (#2785)

GnomedDev committed:
Use ToArrayString for int -> str (#2786)

cheesycod committed:
Rename `unstable_discord_api` to `unstable` and disable it by default (#2789)

This also disables the `#[non_exhaustive]` attribute on `FullEvent` if
`unstable` is enabled as a feature.

It is already expected that any change to Serenity with `unstable` enabled may
or may not break code using this feature and making this enum exhaustive allows
for use cases such as fine-grained audit logging, where new variants should be
a compiler error.

Co-authored-by: Gnome! <david2005thomas@Gmail.com>

GnomedDev committed:
Remove deprecated Channel::is_nsfw methods (#2792)

GnomedDev committed:
Replace manual Hash/Eq implementation on User/Role (#2793)

GnomedDev committed:
Remove useless method wrappers (#2795)

GnomedDev committed:
Fix unused variable warning from rebase

GnomedDev committed:
Remove deprecated mapped refs and Cache::channels (#2796)

GnomedDev committed:
Use ExtractMap instead of HashMap when possible (#2797)

GnomedDev committed:
Remove examples and inaccurate docs from HTTP (#2805)

GnomedDev committed:
Use NonMax in HTTP function signatures (#2803)

arqunis committed:
Fix infinite loop in event deserialisation

GnomedDev committed:
Make Guild::shard_id follow additive features (#2813)

This PR:
- Removes the `cache` feature locked function that called shard_count
for the user, since that is useless and breaks the "all features should
be additive" guidance
- Removes the useless example that just showed how to call a function
- Removed duplicate documentation on Guild, PartialGuild, and
InviteGuild just to direct to the canonical documentation where the
function is implemented.

GnomedDev committed:
Remove deprecated `Member` and `Guild` methods (#2817)

Followup to #2816

---------

Co-authored-by: Alex M. M. <acdenissk69@gmail.com>

GnomedDev committed:
Replace `CacheHttp` with `&Http` in more methods (#2818)

This is possible now more methods are cache-less.

GnomedDev committed:
Remove `serde_json::Value` from `Http` (#2806)

This avoids allocating instances of `serde_json::Value`s.

GnomedDev committed:
Add audit log reasons to remaining model methods (#2821)

GnomedDev committed:
Remove even more impl CacheHttp (#2822)

These were found by auditing .http() calls, instead of just manually
looking at `impl CacheHttp`.

UserIsntAvailable committed:
Remove `fs` feature for `tokio` by default (#2824)

I currently working with wasm on one of my personal projects. I'm would
like to use `serenity` to get access to the discord API response models,
but currently the library can not the compiled to wasm, because the
`tokio/fs` feature is not supported in that target.

I went ahead and removed the need to include that feature by default. By
the looks of it, it only was used with the `builder` and `model`
features, so I conditionally added `tokio/fs` back when those features
are being used (the `model` feature already depends on `builder` so that
doesn't need to be changed).

GnomedDev committed:
Update `reqwest` to v0.12 (#2826)

This has to be on the next branch as reqwest is exposed in a couple
places.

GnomedDev committed:
Only include `cache::Settings::time_to_live` if temp cache is enabled (#2833)

GnomedDev committed:
Add resume_gateway_url support (#2832)

This announcement was missed (around August 2022) and probably the cause
of a lot of the disconnections.

jamesbt365 committed:
Correct documentation for `UserId::to_user` (#2809)

The documentation was never updated after #2662 removed the user cache.

NotNorom committed:
Use CreateAttachment for EditGuild::banner (#2844)

GnomedDev committed:
Fix AuditLogEntry::user_id (#2846)

This field is documented as nullable but that isn't reflected in the
model. I also took the opportunity to replace the Option<Vec> with a
default Vec.

mkrasnitski committed:
Restore match arm for `Error::Gateway` variant (#2847)

This was mistakenly removed in #2278 and wasn't caught because of the
wildcard pattern in the match.

GnomedDev committed:
Remove multiple event handler support (#2849)

TheCataliasTNT2k committed:
Use RwLock for active collectors collection (#2851)

TheCataliasTNT2k committed:
Adds Framework::dispatch_automatically
Makes client::dispatch::update_cache_with_event public, so it can be called manually by developers

jamesbt365 committed:
Provide old presence data from the cache on presence update (#2852)

TheCataliasTNT2k committed:
Allow filtering events before they are dispatched (#2853)

GnomedDev reviewed (changes requested):
I dislike `update_cache_with_event` becoming public, as it's quite internal. If you could make a nicer API for this that doesn't require making internal modules public, that would resolve this.

TheCataliasTNT2k commented:
The only thing in this method I really care about, is to get the returned values.
I do not really want to copy this function to get the FullEvent objects to use them in my library. If this can be achieved in any other way (while not losing the cache functionality) I am happy.

cheesycod commented:
This would be useful for me as well

jamesbt365 committed:
Remove permission checks (#2855)

Users should realistically be checking the permissions themselves, or
handling the HTTP error from Discord.

This removes any cases where permission checking inside the library is
broken because of Discord's changes or due to oversights.

This also changes the documentation on the prune functionality, this was
recently changed to also require `MANAGE_GUILDS` as well as
`KICK_MEMBERS`.

GnomedDev committed:
Implement ToArrayString for Mention (#2857)

cheesycod committed:
Add PartialGuild::greater_member_hierarchy + member_highest_role (#2856)

GnomedDev committed:
Update bool_to_bitflags to fix compilation without unstable

GnomedDev committed:
Remove recently deprecated items

jamesbt365 committed:
Remove RUSTFLAGS from .cargo/config.toml (#2875)

This isn't needed anymore because simd-json has been removed on `next`.

GnomedDev committed:
Remove deprecated link_ensured method (#2898)

TheCataliasTNT2k commented:
I can not really create a better API, because I need to update the cache.
Do you have any suggestions?
(I will resolve the conflicts when all changes are approved.)

> I dislike `update_cache_with_event` becoming public, as it's quite internal. If you could make a nicer API for this that doesn't require making internal modules public, that would resolve this.

Already did the rebase (or rather just moved my commits to the new HEAD).

GnomedDev committed:
Add GuildId param to ChannelId::to_channel (#2895)

GnomedDev committed:
Clean up `Message::reply` (#2897)

- Gets rid of duplicate documentation
- Removes unnecessary monomorphisation of the `Cow<'_, str>` argument
- Removes the useless `reply_mention` method
- Respects the default_allowed_mentions.

GnomedDev committed:
Use `ArrayString` in more places via `aformat!` (#2902)

This also bumps `to-arraystring` up to 0.2, required for `aformat`.

cheesycod commented:
Any updates on this?

TheCataliasTNT2k commented:
I don't think so 🤷
I am just using a fork which has everything I need, which is rejected by the maintainers.

mkrasnitski committed:
Bump MSRV to Rust 1.79 (#2916)

The actual version change was made in #2902, but left README.md
unchanged.

mkrasnitski committed:
Remove deprecated `CreateInteractionResponse::PremiumRequired` variant (#2928)

This was deprecated on `current` in #2905, so it can be removed here on
`next`.

valentinegb committed:
Add `OAuth2Token`, `OAuth2TokenRevocation`, and `OAuth2AuthorizationCurrent` to `Route` (#2935)

This adds routes to `Route` that are related to OAuth2 and token exchanges.

This is a step towards resolving #762 and #2475.

tazz4843 committed:
Allow setting both normal and raw event handlers (#2933)

mkrasnitski committed:
Merge `client` feature/module into `gateway` (#2947)

The `client` and `gateway` features were very tightly coupled so it
makes the most sense to just merge them because there's no real path
towards separating them ergonomically.

A few breaking changes:
* Removed `ClientError` because it was unused.
* The `client` module is moved to be a submodule under the `gateway`
module.
* Moved sharding-specific gateway types into their own `sharding`
submodule for clarity's sake, but re-exported them.

I also adjusted the gateway documentation to hopefully make it as clear as
possible.

Note that the `Cargo.toml` changes are mostly superficial - in
particular, changing `gateway` to depend on `model` (instead of `http`
like how `client` did) isn't actually much of a change because trying to
use `gateway` with only a subset of `model` enabled will fail to
compile, and that was also the case for `client` previously.

mkrasnitski committed:
Remove methods on `ClientBulder` that don't serve a purpose (#2953)

The `ClientBuilder` struct is gated behind the `gateway` feature, which
means that the client will receive the current application ID over the
gateway when first authenticating. Therefore it doesn't make sense to
set the application ID through the `ClientBuilder` as it will get
overrided anyway.

Also, the `get_token` method is pretty useless because the only source
of the token would be whatever the user would have passed to
`ClientBuilder::new`, meaning they would already have a way to read out
its value.

mkrasnitski committed:
Fix and make `edit_role_position` plural (#2954)

GnomedDev committed:
Fix model functions using `SerializeIter` (#2955)

When passing `SerializeIter` in from outside, it counts as passing `!Send`
over an await point and therefore making these functions unusable.

This fixes that by moving the `SerializeIter` usage inside of `Http`, or
replacing with a slice when necessary.

mkrasnitski committed:
Remove `InternalEventHandler` (#2952)

Since having both a regular event handler and a raw handler at the same time is
now supported, the `InternalEventHandler` enum is no longer needed to ensure
mutual exclusivity. 

Additionally, there was a subtle bug introduced when the `Both` variant was
added: ratelimit events and shard stage update events would not be dispatched
because the conditionals to dispatch those events only checked for the `Normal`
variant. Inlining the regular/raw event handler fields improves ergonomics and
fixes said bug.

ivinjabraham committed:
Truncate cache if `max_messages` is set to a lower value at runtime (#2957)

Fixes #2884

ivinjabraham committed:
Add enum for JSON error codes in `DiscordJsonError` (#2964)

Fixes #2752

GnomedDev committed:
Update to DashMap 6 (#2961)

mkrasnitski committed:
Fix unused warnings when compiling without certain features (#2970)

This is mostly just clean up, moving things around and gating things
behind features so that unused warnings in CI go away.

mkrasnitski committed:
Fix public doctests that are marked `ignore` (#2976)

Doctests on public items should be maintained accurate as the API
changes, so having them be type-checked makes the most sense.

I removed the doctest for the `CacheUpdate` trait as it was really 
out of date and not really possible to fix up.

TapGhoul committed:
Allow embeds to be optional in followup messages (#2968)

GnomedDev committed:
Swap `Vec<T>` for `Cow<[T]>` in `CreateActionRow::Buttons` (#2985)

Closes #2984.

GnomedDev committed:
Remove more `json!` usage (#2986)

GnomedDev committed:
Replace outdated secrecy library with own impl (#2997)

`secrecy` has had major changes in it's transition from `0.8` -> `0.10`
that would make it even less useful than it is now.

This just brings the basic implementation in-house and exposes it
publicly, which is somehow less lines of code that trying to use the
`0.8` API, let alone `0.10`.

---------

Co-authored-by: jamesbt365 <jamesbt365@gmail.com>

GnomedDev committed:
Check for lints which change public API (#2999)

This is fine to do on next, and fixed a couple things. I also swapped
the `allow`s for `expect`s because why not.

GnomedDev committed:
Stop using serde_json::Value in AutocompleteChoice::value (#3000)

More type safe, more performant, less reliant on serde_json, what's not
to love!

cheesycod committed:
feat: Make FullEvent serializable (#3010)

GnomedDev committed:
Add `Context::data_ref` (#3017)

This is needed for use cases like returning a borrow into the user data
from a function which takes a reference to Context.

GnomedDev committed:
Delete passthrough methods (#3023)

This deletes methods which simply pass through to their `id` fields, as
they are:
- Misleading, as a user may do an HTTP call or cache lookup to get a
model just to use it's ID
- Lead to a bunch of maintenance issues when having 2..=4 copies of
every method.
- Hurt compile times, as rustc is having to check signatures, calls,
construct async state machines, and maybe even inline to generate code
multiple times.

For methods which simply only used their `id` field but did not have a
version on their respective ID, they were moved to their ID.

This doesn't have a corresponding deprecation PR to current, as current
still has permission checks and other "helpers" which rely on the model.

GnomedDev committed:
Swap CreateAttachment::data to Bytes (#3016)

This exposes the `bytes` dependency but in turn allows for people to
pass responses directly from reqwest to discord without cloning the
data, as well as allowing reuploading cached data without cloning or
leaking to get a static reference.

The bytes dependency also has to be made non-optional, but this is fine
as it was already due to `aformat` depending on it via `bytestring`.

mkrasnitski committed:
Add `EditCommand` builder (#3028)

This separates the the builders for creating versus editing a command,
since it's not possible to change the `type` of a command (in serenity
this is the `kind` field). Also, the `name` field is not required when
editing a command.

GnomedDev committed:
Remove deprecated GuildChannel::permissions_for_user (#3033)

mkrasnitski committed:
Improve some docs (#3027)

Improved wording and removed some sentences that are no longer relevant.
These docs are mostly the ones that were touched by #2988.

cheesycod committed:
Add config for the shard start wait time (#3031)

Co-authored-by: Gnome! <david2005thomas@Gmail.com>

GnomedDev committed:
Remove Option around Interaction::app_permissions (#3034)

Discord documents this as always present, even with Ping interactions.

GnomedDev committed:
Remove deprecated permissions methods (#3042)

This follows up on #3037, and also removes the now unnecessary Option
around the GuildChannel param.

DPlayer234 committed:
Update `tokio-tungstenite` dependency to 0.24 (#3048)

GnomedDev committed:
Perform some misc TODO cleanups (#3050)

fgardt committed:
Transport compression support (#3036)

Co-authored-by: GnomedDev <david2005thomas@gmail.com>

GnomedDev committed:
Get rid of redundant status argument in Http::wind (#3049)

Turns out this was always being passed 204, except one case where it
should have been deserialising the `Member` return type anyway. This
fixes both issues.

mkrasnitski committed:
Perform some minor cleanups (#3051)

mkrasnitski committed:
Improve token semantics and clean up gateway feature gates (#3052)

GnomedDev committed:
Add Token to serenity::all (#3054)

This allows it to imported via `poise::serenity_prelude` which is often
`use`'d as `serenity`.

arqunis committed:
Use correct token prefix after parsing

jamesbt365 committed:
Add guild_id to ReactionRemoveAll (#3058)

This was missing on FullEvent but was present within the event itself.

GnomedDev committed:
Bump the discriminant of audit_log::Action to u16 (#3062)

Values are getting quite close to the u8 limit (193), so let's
preemptively bump this to avoid future breaking changes.

fgardt committed:
Fix transport compression feature & add CI tests (#3067)

mkrasnitski committed:
Factor out collector methods on models into traits (#3055)

Moves the feature-gated collector methods on `UserId`, `MessageId`, etc.
into four traits:
- `CollectMessages` 
- `CollectReactions`
- `CollectModalInteractions` 
- `CollectComponentInteractions`

This also moves the quick modal machinery into the collector module and defines
a `QuickModal` trait. 

This fully removes any collector feature gates from the model types.

mkrasnitski committed:
Remove unused `json.rs` file (#3071)

This file was removed from the module tree but never actually deleted.

mkrasnitski committed:
Fix clippy warning on next (#3078)

Fixes CI failure

TheCataliasTNT2k committed:
Add a function for parsing a channel ID from a channel URL (#3075)

Similar to message URLs, Discord also provides URLs for guild channels.

Additionally, the function is added as an alternative for parsing a `Channel` from a string. 
Private channels are not affected by this change.

mkrasnitski committed:
Fix clippy warning (#3081)

jamesbt365 committed:
Add `newly_created` bool field to the `ThreadCreate` event (#3084)

https://discord.com/developers/docs/events/gateway-events#thread-create

arqunis committed:
Fix deserialisation of `Token` type (#3086)

The `Deserialize` implementation neglects to add the `Bot ` prefix to
the string when it is deserialised.

This adds `TryFrom` implementations for `&str` and `String` and tells
serde to deserialise `Token` using the `TryFrom<&str>` implementation,
which will prepend the `Bot ` prefix.

Fixes #3085

arqunis committed:
Fix issues from rebase

arqunis committed:
Remove `ContentSafeOptions::show_discriminator`

GnomedDev committed:
Change MessageUpdateEvent to just contain a full Message (#3061)

mkrasnitski committed:
Improve gateway connection/resume logic (#3099)

This commit refactors how the gateway connection being closed gets handled,
and also reworks how resuming is performed. If a resume fails, or if the
session id is invalid/doesn't exist, the shard will fall back to restart
+ reidentify after a 1 second delay. This behavior was only present in
some circumstances before.

Also, cleaned up the loop in `ShardRunner::run` by adding a
`ShardAction::Dispatch` variant, since event dispatch was already
mutually exclusive to hearbeating, identifying, and restarting. The
overall effect is less interleaving of control flow.

Plus, removed the `Shard::{reconnect, reset}` functions as they were
unused.

A notable change is that 4006 is no longer considered a valid close code
as it is undocumented, and neither is 1000, which tungstenite assigns as
`Normal` or "clean". We should stick to the [table of close
codes](https://discord.com/developers/docs/topics/opcodes-and-status-codes#gateway-gateway-close-event-codes)
provided by Discord.

Cabbache committed:
Update `tokio-tungstenite` to v0.26.1

arqunis committed:
Fix clippy

mkrasnitski committed:
Fix disconnections on successful gateway resume (#3102)

A regression introduced by #3099 was that successful resumes will break
out of the loop inside `ShardRunner::run`, but they shouldn't (or
rather, didn't before). Therefore, only break out of the loop if the
resume failed and we had to fallback to reidentifying.

Freyja-moth committed:
Deserialise `Token` from `String`s instead of `&str`s (#3104)

mkrasnitski committed:
Fix clippy

mkrasnitski committed:
Fix compilation when the `tracing_instrument` feature is enabled (#3110)

xacrimon committed:
Swap `fxhash` for `foldhash` due to hash quality issues (#3111)

As the title notes, this commit replaces fxhash for foldhash as used in the
cache. dashmap, due to it's sharding, has to share entropy with what's handed
down to internal maps. Since `hashbrown` and by extension `std` use various
sections of the high bit range for special grouping & sorting, dashmap is left
with the only option to shard on low bits.

This, however, presents problems, because fxhash outputs hashes of very bad
quality, with only the high bits having any real entropy. This was probably a
solid choice back in 2018 when we lacked other good fast alternatives. But
since then `ahash` matured and we've had significant research and development
in "good enough" hashing for datastructures with short keys, [the most recent
step forward coming from a rather well known face][foldhash]. This improves
shard selection quite a bit and reduces contention significantly. Using fxhash
in a dashmap specific benchmark causes contention to go up by 3-8x when keys
are k-sortable with time (Discord snowflakes) on an M1 Pro.

[foldhash]: https://github.com/orlp/foldhash

cheesycod committed:
Replace Delete-Messages-Days with Delete-Messages-Seconds (#3088)

This replaces the time unit that determines how many messages in a
period of X time should be deleted when a user is banned. Discord
expresses this time in seconds, whereas Serenity exposed it in days. The
limit is still 7 days, but with this, users of Serenity can be more
precise with which messages they want to delete.

mkrasnitski committed:
Remove deprecated field inside `GatewayEvent::Heartbeat` (#3115)

xacrimon committed:
Avoid temporarily deserializing gateway messages to an untyped enum+map tree (#3114)

Avoid temporarily deserializing gateway messages to a `serde_json::Map<String,
Value>` before typed deserialization to an `Event`. 

Previously, this meant the creation of a `serde_json::Value` tree, causing
creation and immediately after the destruction of upwards of hundreds to
thousands of owned strings and btreemaps for every handled gateway event.

GnomedDev commented:
Closing as you have said you are not interested in getting this to a mergable state.

TheCataliasTNT2k commented:
> Closing as you have said you are not interested in getting this to a mergable state.

I did not say, that I am not interested.
I proposed changes, you were unhappy, and I asked you, which solution you would prefer to achieve the same functionality [here](https://github.com/serenity-rs/serenity/pull/2854#issuecomment-2086969687) and [here](https://github.com/serenity-rs/serenity/pull/2854#issuecomment-2159149613).
You never replied to any of these comments.

Give me a solution, which you would accept, I will implement it and rebase my branch :woman_shrugging: .

TheCataliasTNT2k commented:
If you need a reference what I am doing with this change:
I created my own framework based off of poise and serenity.
I want to be able to control event handling how I need it to to be.
So I use the "raw_event_handler" method provided by poise / serenity and handle the events myself.
But I need a way to convert the raw events into FullEvents and update the cache.
This is done by this function exactly and therefore it needs to be public.
There is no "nicer API" I can think of.
